components/content_settings/core/browser/brave_cookie_settings.cc

```diff
@@ -0,0 +1,120 @@
+#include "url/gurl.h"
+
+namespace {
+
+bool IsValidSetting(ContentSetting setting) {
+  return (setting == CONTENT_SETTING_ALLOW ||
+          setting == CONTENT_SETTING_SESSION_ONLY ||
+          setting == CONTENT_SETTING_BLOCK);
+}
+
+bool IsAllowed(ContentSetting setting) {
+  DCHECK(IsValidSetting(setting));
+  return (setting == CONTENT_SETTING_ALLOW ||
+          setting == CONTENT_SETTING_SESSION_ONLY);
+}
+
+}  // namespace
+
+namespace content_settings {
+
+using namespace net::registry_controlled_domains;
+
+BraveCookieSettings::BraveCookieSettings(
+    HostContentSettingsMap* host_content_settings_map,
+    PrefService* prefs,
+    const char* extension_scheme)
+    : CookieSettings(host_content_settings_map, prefs, extension_scheme)
+{ }
+
+BraveCookieSettings::~BraveCookieSettings() { }
+
+void BraveCookieSettings::GetCookieSetting(const GURL& url,
+    const GURL& first_party_url,
+    content_settings::SettingSource* source,
+    ContentSetting* cookie_setting) const {
+  GetCookieSetting(url,
+    first_party_url,
+    first_party_url,
+    source,
+    cookie_setting);
+}
+
+void BraveCookieSettings::GetCookieSetting(const GURL& url,
+    const GURL& first_party_url,
+    const GURL& tab_url,
+    content_settings::SettingSource* source,
+    ContentSetting* cookie_setting) const {
+  DCHECK(cookie_setting);
+  CookieSettings::GetCookieSetting(url, first_party_url, source,
+      cookie_setting);
+
+  if (*cookie_setting == CONTENT_SETTING_BLOCK) {
+    *cookie_setting = CONTENT_SETTING_BLOCK;
+    return;
+  }
+
+  GURL primary_brave_url = tab_url == GURL("about:blank") ?
+      first_party_url : tab_url;
+
+  // Check the Brave shields setting, if it is off, just return without
+  // blocking anything.
+  ContentSetting brave_shields_setting =
+      host_content_settings_map_->GetContentSetting(
+          primary_brave_url, GURL(),
+          CONTENT_SETTINGS_TYPE_PLUGINS, brave_shields::kBraveShields);
+  if (brave_shields_setting == CONTENT_SETTING_BLOCK) {
+    *cookie_setting = CONTENT_SETTING_ALLOW;
+    return;
+  }
+
+  // First party setting of block means always block everything
+  ContentSetting brave_1p_setting = host_content_settings_map_->GetContentSetting(
+      primary_brave_url, GURL("https://firstParty/"),
+      CONTENT_SETTINGS_TYPE_PLUGINS, brave_shields::kCookies);
+  if (brave_1p_setting == CONTENT_SETTING_BLOCK) {
+    *cookie_setting = CONTENT_SETTING_BLOCK;
+    return;
+  }
+
+  // Third party setting of allow, means always allow everything
+  ContentSetting brave_3p_setting =
+      host_content_settings_map_->GetContentSetting(
+          primary_brave_url, GURL(),
+          CONTENT_SETTINGS_TYPE_PLUGINS, brave_shields::kCookies);
+  if (brave_3p_setting == CONTENT_SETTING_ALLOW) {
+    *cookie_setting = CONTENT_SETTING_ALLOW;
+    return;
+  }
+
+  // Otherwise determine based on if the cookie is third-party or not
+  if (brave_3p_setting != CONTENT_SETTING_ALLOW &&
```

Comment:
few thoughts regarding to this if block
1) I think brave_3p_setting != CONTENT_SETTING_ALLOW is always true here, could be removed
2) why is this block has nested if conditions, instead of just if first-party or not?
3) early return in else is redundant

Reply:
1. agreed.
2. we only want to enter the block if `first_party_url` is empty.  We might want to return brave_1p_setting in that case though, I think that's the case but I'll check.
3. will remove.